presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/ImplementBernoulliSampleAsFilter.java

```diff
@@ -49,10 +50,18 @@
 public class ImplementBernoulliSampleAsFilter
         implements Rule
 {
+    private static final Pattern PATTERN = Pattern.node(SampleNode.class);
+
+    @Override
+    public Pattern getPattern()
+    {
+        return PATTERN;
+    }
+
     @Override
     public Optional<PlanNode> apply(PlanNode node, Lookup lookup, PlanNodeIdAllocator idAllocator, SymbolAllocator symbolAllocator, Session session)
     {
-        if (!(node instanceof SampleNode)) {
+        if (!PATTERN.matches(node)) {
```

Comment:
Since all the rules specify their patterns, it's the IterativeOptimizer that should to the match on the pattern, using `Rule#getPattern()`. This will remove quite a lot of repetition and will prevent bugs. A possible bug because of this duplication would be: a rule is fired for a complex pattern like "ProjectNode with at least two projections" - of which only the node type will be pre-matched by the `RuleStore`

Alternatively, `getPattern()` can be made part of an adapter interface, say `PatternMatchingRule`, as described in the [Pattern Matching design doc](https://docs.google.com/document/d/17N7CAIwdlbW_kD641ceCJWqeeNJrdPJc1ZrAiz49iuQ/edit#heading=h.ms9aksjvfb8x).

You mentioned 'this is a design decision'. I think this is the moment to take it, else we'll introduce bugs for the devs extending this mechanism.